### PR TITLE
hyper-v enlightenment: Install epel repo on RHEL guest

### PR DESCRIPTION
**Pull request by peixiu**
cpuid package is included in epel repo, so prepare epel repo for these cases.

1] hv_avic
2] hv_flag_cpuid_check

ID: 1765
Signed-off-by: peixiu <phou@redhat.com>

### CODE REVIEW

**Comment by peixiu** [comment]
python3 ConfigTest.py --testcase=hv_flag_cpuid_check --guestname=RHEL.9.4 --platform=x86_64 --clone=yes --machines=q35

 (1/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1028.91 s)
 (2/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_ext.q35: STARTED
 (2/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_ext.q35: PASS (308.22 s)
 (3/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_direct.q35: STARTED
 (3/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_direct.q35: PASS (187.84 s)

Check the debug log, the result as expected:
[stdlog] 2023-12-26 05:42:28,195 aexpect.client DEBUG| [10.73.74.171] Sending command: rpm -qa | grep cpuid
[stdlog] 2023-12-26 05:42:32,523 aexpect.client DEBUG| [10.73.74.171] Sending command: echo $?
[stdlog] 2023-12-26 05:42:32,625 aexpect.client DEBUG| [10.73.74.171] Sending command (safe): yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm
[stdlog] 2023-12-26 05:42:41,758 aexpect.client DEBUG| [10.73.74.171] Sending command: yum -y install cpuid
[stdlog] 2023-12-26 05:43:01,291 aexpect.client DEBUG| [10.73.74.171] Sending command: echo $?
[stdlog] 2023-12-26 05:43:01,393 avocado.test INFO | Context: Check the corresponding CPUID entries with the flag hv_tlbflush_ext
[stdlog] 2023-12-26 05:43:01,393 aexpect.client DEBUG| [10.73.74.171] Sending command: cpuid -r -1 -l 0x40000003

**Comment by peixiu** [comment]
@yanan-fu Could you help to review it? thanks~

**Comment by yanan-fu** [comment]
@nickzhq  could you please help review ?  It use the same solution with case hv_enforce_cpuid_msr_check to setup the epel repo.

**Comment by nickzhq** [comment]
> @nickzhq could you please help review ? It use the same solution with case hv_enforce_cpuid_msr_check to setup the epel repo.

ACK. 
cc @yanan-fu

File: qemu/tests/cfg/hv_avic.cfg [hunk 1]
```diff
@@ -5,5 +5,7 @@
     required_qemu = [6.2.0,)
     cpu_model_flags += ",hv_avic"
     cpuid_chk_cmd = "rpm -qa | grep cpuid"
-    cpuid_pkg = ""
+    cpuid_pkg = "cpuid"
+    RHEL.9:
+        repo_install_cmd = "yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm"
```
Comment (yanan-fu) [inline qemu/tests/cfg/hv_avic.cfg:10]: The limited QEMU version - QEMU 6.2 is used by RHEL8.6.  
Setting the epel repo for RHEL.9 is not enough,   and it is not friendly for the new product version.

Comment (yanan-fu) [reply qemu/tests/cfg/hv_avic.cfg:10]: Assume the fedora epel repository naming rule is consistent:
`epel-release-latest-$RHEL_X_VERSION.noarch.rpm`.

Suggest replace the RHEL_X_VERSION during case running.

**Comment by peixiu** [comment]
python3 ConfigTest.py --testcase=hv_avic --guestname=RHEL.8.10.0 --platform=x86_64 --clone=no --machines=q35

 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.10.0.x86_64.io-github-autotest-qemu.hv_avic.q35: STARTED
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.10.0.x86_64.io-github-autotest-qemu.hv_avic.q35: PASS (66.29 s)

Check logs: 
[stdlog] 2024-01-05 05:29:15,933 aexpect.client DEBUG| [10.72.142.6] Sending command (safe): **yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm**
[stdlog] 2024-01-05 05:29:23,560 aexpect.client DEBUG| [10.72.142.6] Sending command: yum -y install cpuid
[stdlog] 2024-01-05 05:29:27,455 aexpect.client DEBUG| [10.72.142.6] Sending command: echo $?
[stdlog] 2024-01-05 05:29:27,556 avocado.test INFO | Context: Check the corresponding CPUID entries with the flag 'hv-avic'

 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_ext.q35: STARTED
 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_ext.q35: PASS (153.84 s)
 (2/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_direct.q35: STARTED
 (2/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_direct.q35: PASS (142.77 s)

[stdlog] 2024-01-05 05:43:15,756 aexpect.client DEBUG| [10.72.142.141] Sending command (safe): **yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm**
[stdlog] 2024-01-05 05:43:22,772 aexpect.client DEBUG| [10.72.142.141] Sending command: yum -y install cpuid
[stdlog] 2024-01-05 05:43:26,229 aexpect.client DEBUG| [10.72.142.141] Sending command: echo $?
[stdlog] 2024-01-05 05:43:26,330 avocado.test INFO | Context: Check the corresponding CPUID entries with the flag hv_tlbflush_ext

**Comment by peixiu** [comment]
@yanan-fu Tested with the latest patch, result as expected, I pasted some logs for that.
Please help to review it again, thanks~

File: qemu/tests/cfg/hv_avic.cfg [hunk 2]
```diff
@@ -5,5 +5,6 @@
     required_qemu = [6.2.0,)
     cpu_model_flags += ",hv_avic"
     cpuid_chk_cmd = "rpm -qa | grep cpuid"
-    cpuid_pkg = ""
+    cpuid_pkg = "cpuid"
+    repo_install_cmd = "yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-%s.noarch.rpm"
```
Comment (yanan-fu) [inline qemu/tests/cfg/hv_avic.cfg:9]: Suggest:
```
RHEL:
    repo_install_cmd = "dnf install -y ....."
```

refer: https://docs.fedoraproject.org/en-US/epel/

File: qemu/tests/hv_avic.py [hunk 3]
```diff
@@ -13,6 +14,27 @@ def run(test, params, env):
     param params: the test params
     param env: the test env object
     """
+
+
+    def _get_rhel_main_ver():
+        """
+        Get host main version.
+        """
+        cmd = "cat /etc/redhat-release | awk '{print $6}' | awk -F'.' '{print $1}'"
```
Comment (yanan-fu) [inline qemu/tests/hv_avic.py:23]: Suggest using the `/etc/os-release` instead,  it is a standard file used by many Linux distributions.

File: qemu/tests/hv_flag_cpuid_check.py [hunk 4]
```diff
@@ -50,6 +51,26 @@ def run_cpuid_check(check_register, check_bit):
...
+
+
+    def install_epel_repo():
+        """
+        Install required packages based on RHEL version.
+        """
+        rhel_main_ver = _get_rhel_main_ver()
+        repo_install_cmd = params.get("repo_install_cmd") % rhel_main_ver
+        session.cmd_output_safe(repo_install_cmd)
+        time.sleep(5)
```
Comment (yanan-fu) [inline qemu/tests/hv_flag_cpuid_check.py:71]: As the repo_install_cmd is not for all distribution, suggest:
```
repo_install_cmd = params.get("repo_install_cmd")
if not repo_install_cmd:
    return
repo_install_cmd = repo_install_cmd % rhel_major_ver
...
...
```

File: qemu/tests/hv_avic.py [hunk 5]
```diff
@@ -13,6 +14,27 @@ def run(test, params, env):
...
+        """
+        cmd = "awk '/PRETTY_NAME/ {print}' /etc/os-release | awk '{print $5}' | awk -F '.' '{print $1}'"
+        rhel_main_ver = session.cmd_output(cmd)
+        return rhel_main_ver.strip()
+
+    def install_epel_repo():
+        """
+        Install required packages based on RHEL version.
+        """
+        rhel_main_ver = _get_rhel_main_ver()
```
Comment (yanan-fu) [inline qemu/tests/hv_avic.py:30]: It is better to call it after repo_install_cmd check. others LGTM

**Comment by peixiu** [comment]
@yanan-fu Updated, thanks~

Tested with the latest code:

- python3 ConfigTest.py --testcase=hv_flag_cpuid_check --guestname=RHEL.9.4 --platform=x86_64 --clone=yes --machines=q35

 (1/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1084.45 s)
 (2/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_ext.q35: STARTED
 (2/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_ext.q35: PASS (236.42 s)
 (3/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_direct.q35: STARTED
 (3/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_flag_cpuid_check.hv_tlbflush_direct.q35: PASS (186.42 s)

- python3 ConfigTest.py --testcase=hv_avic --guestname=RHEL.9.4 --platform=x86_64 --clone=yes --machines=q35

 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1078.17 s)
 (2/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_avic.q35: STARTED
 (2/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hv_avic.q35: PASS (185.52 s)

**Review by yanan-fu — APPROVED** [review]
Ack

**Comment by yanan-fu** [comment]
@nickzhq  could you please help review the latest version, thanks!

**Comment by nickzhq** [comment]
> @nickzhq could you please help review the latest version, thanks!

Hello @peixiu , I believe we don't need 2 commits at here, can you make just one commit? Thanks!

**Comment by peixiu** [comment]
> > @nickzhq could you please help review the latest version, thanks!
> 
> Hello @peixiu , I believe we don't need 2 commits at here, can you make just one commit? Thanks!

Thanks comment, updated it~

**Comment by nickzhq** [comment]
> > > @nickzhq could you please help review the latest version, thanks!
> > 
> > 
> > Hello @peixiu , I believe we don't need 2 commits at here, can you make just one commit? Thanks!
> 
> Thanks comment, updated it~

ACK. 
Thanks!

**Review by nickzhq — APPROVED** [review]

**Comment by yanan-fu** [comment]
Ack

**Pull request merged**